File: Fanficly/Networking/AO3Endpoints.swift
import Foundation

enum AO3Endpoints {
    static func search(filters: AO3SearchFilters, page: Int, base: URL) throws -> URL {
        var components = URLComponents(url: base.appending(path: "/works/search"), resolvingAgainstBaseURL: false)
        var items = filters.queryItems()
        if page > 1 { items.append(URLQueryItem(name: "page", value: String(page))) }
        components?.queryItems = items
        guard let url = components?.url else { throw AO3Error.parseFailed(reason: "Bad search URL") }
        return url
    }

    static func work(id: Int, base: URL) throws -> URL {
        var components = URLComponents(url: base.appending(path: "/works/\(id)"), resolvingAgainstBaseURL: false)
        components?.queryItems = [
            URLQueryItem(name: "view_full_work", value: "true"),
            URLQueryItem(name: "view_adult", value: "true"),
        ]
        guard let url = components?.url else { throw AO3Error.parseFailed(reason: "Bad work URL") }
        return url
    }

    static func epub(workId: Int, base: URL) throws -> URL {
        base.appending(path: "/downloads/\(workId)/work.epub")
    }

    static func download(workId: Int, format: WorkExportFormat, base: URL) throws -> URL {
        base.appending(path: "/downloads/\(workId)/work.\(format.ext)")
    }

    static func authorWorks(name: String, page: Int, base: URL) throws -> URL {
        var components = URLComponents(url: base.appending(path: "/users/\(name)/works"),
                                       resolvingAgainstBaseURL: false)
        if page > 1 { components?.queryItems = [URLQueryItem(name: "page", value: String(page))] }
        guard let url = components?.url else { throw AO3Error.parseFailed(reason: "Bad author works URL") }
        return url
    }

    static func userBookmarks(name: String, page: Int = 1, base: URL) throws -> URL {
        var components = URLComponents(url: base.appending(path: "/users/\(name)/bookmarks"),
                                       resolvingAgainstBaseURL: false)
        if page > 1 { components?.queryItems = [URLQueryItem(name: "page", value: String(page))] }
        guard let url = components?.url else { throw AO3Error.parseFailed(reason: "Bad bookmarks URL") }
        return url
    }

    static func userSubscriptions(name: String, page: Int = 1, base: URL) throws -> URL {
        var components = URLComponents(url: base.appending(path: "/users/\(name)/subscriptions"),
                                       resolvingAgainstBaseURL: false)
        if page > 1 { components?.queryItems = [URLQueryItem(name: "page", value: String(page))] }
        guard let url = components?.url else { throw AO3Error.parseFailed(reason: "Bad subscriptions URL") }
        return url
    }

    static func login(base: URL) throws -> URL {
        base.appending(path: "/users/login")
    }

    static func workComments(id: Int, base: URL) throws -> URL {
        var components = URLComponents(url: base.appending(path: "/works/\(id)"), resolvingAgainstBaseURL: false)
        components?.queryItems = [
            URLQueryItem(name: "view_adult", value: "true"),
            URLQueryItem(name: "show_comments", value: "true"),
        ]
        components?.fragment = "comments"
        guard let url = components?.url else { throw AO3Error.parseFailed(reason: "Bad comments URL") }
        return url
    }

    /// A single chapter's page with its comment thread shown — AO3 threads
    /// comments per chapter, so this is what we read/post against when the
    /// reader knows which chapter id it's on.
    static func chapterComments(workId: Int, chapterId: Int, base: URL) throws -> URL {
        var components = URLComponents(url: base.appending(path: "/works/\(workId)/chapters/\(chapterId)"),
                                       resolvingAgainstBaseURL: false)
        components?.queryItems = [
            URLQueryItem(name: "view_adult", value: "true"),
            URLQueryItem(name: "show_comments", value: "true"),
        ]
        components?.fragment = "comments"
        guard let url = components?.url else { throw AO3Error.parseFailed(reason: "Bad chapter comments URL") }
        return url
    }

    static func workSubscriptions(id: Int, base: URL) throws -> URL {
        base.appending(path: "/works/\(id)/subscriptions")
    }

    static func workCommentsPost(id: Int, base: URL) throws -> URL {
        base.appending(path: "/works/\(id)/comments")
    }

    static func autocomplete(field: String, term: String, base: URL) throws -> URL {
        var components = URLComponents(url: base.appending(path: "/autocomplete/\(field)"), resolvingAgainstBaseURL: false)
        // Percent-encode the term ourselves: URLComponents leaves "/" literal in
        // query values (it's RFC-legal there), but AO3's autocomplete 404s on a
        // raw slash — so ships like "Hermione/Draco" must send "%2F".
        let allowed = CharacterSet.urlQueryAllowed.subtracting(CharacterSet(charactersIn: "/+&=?"))
        let encoded = term.addingPercentEncoding(withAllowedCharacters: allowed) ?? term
        components?.percentEncodedQuery = "term=\(encoded)"
        guard let url = components?.url else { throw AO3Error.parseFailed(reason: "Bad autocomplete URL") }
        return url
    }

    static func mediaFandoms(categoryName: String, base: URL) throws -> URL {
        let encoded = ao3PathEncode(categoryName)
        guard let url = URL(string: "\(base.absoluteString)/media/\(encoded)/fandoms") else {
            throw AO3Error.parseFailed(reason: "Bad media URL")
        }
        return url
    }

    /// A tag's works listing (`/tags/<tag>/works`) — its left filter sidebar
    /// carries the most-used relationships/characters/freeforms with counts.
    static func tagWorks(tagName: String, base: URL) throws -> URL {
        let encoded = ao3PathEncode(tagName)
        guard let url = URL(string: "\(base.absoluteString)/tags/\(encoded)/works") else {
            throw AO3Error.parseFailed(reason: "Bad tag works URL")
        }
        return url
    }

    private static func ao3PathEncode(_ s: String) -> String {
        var result = ""
        for ch in s {
            switch ch {
            case "&": result += "*a*"
            case "/": result += "*s*"
            case ".": result += "*d*"
            case "?": result += "*q*"
            case "'": result += "*at*"
            case " ": result += "%20"
            default:
                if let escaped = String(ch).addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) {
                    result += escaped
                } else {
                    result.append(ch)
                }
            }
        }
        return result
    }
}

File: android/app/src/main/java/io/github/yennster/fanficly/net/AO3Endpoints.kt
package io.github.yennster.fanficly.net

import io.github.yennster.fanficly.model.AO3SearchFilters
import io.github.yennster.fanficly.model.AutocompleteField
import io.github.yennster.fanficly.model.ExportFormat
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl

/**
 * Builds the AO3 URLs the client hits — the Kotlin port of iOS `AO3Endpoints`.
 * OkHttp's [HttpUrl] handles ordinary query encoding; the autocomplete and
 * media-path helpers reproduce the two AO3 encoding quirks the iOS app calls
 * out (slash-in-`term` must be `%2F`; media category path uses AO3's `*a*`/`*s*`
 * scheme).
 */
object AO3Endpoints {
    val BASE: HttpUrl = "https://archiveofourown.org".toHttpUrl()

    fun search(filters: AO3SearchFilters, page: Int): HttpUrl =
        BASE.newBuilder().addPathSegments("works/search").apply {
            filters.queryItems().forEach { (k, v) -> addQueryParameter(k, v) }
            if (page > 1) addQueryParameter("page", page.toString())
        }.build()

    fun work(id: Int): HttpUrl =
        BASE.newBuilder().addPathSegments("works/$id")
            .addQueryParameter("view_full_work", "true")
            .addQueryParameter("view_adult", "true")
            .build()

    fun epub(workId: Int): HttpUrl =
        BASE.newBuilder().addPathSegments("downloads/$workId/work.epub").build()

    fun download(workId: Int, format: ExportFormat): HttpUrl =
        BASE.newBuilder().addPathSegments("downloads/$workId/work.${format.ext}").build()

    fun authorWorks(name: String, page: Int): HttpUrl =
        BASE.newBuilder().addPathSegments("users/$name/works").apply {
            if (page > 1) addQueryParameter("page", page.toString())
        }.build()

    fun userBookmarks(name: String, page: Int = 1): HttpUrl =
        BASE.newBuilder().addPathSegments("users/$name/bookmarks").apply {
            if (page > 1) addQueryParameter("page", page.toString())
        }.build()

    fun login(): HttpUrl = BASE.newBuilder().addPathSegments("users/login").build()

    /** The work page with its comment thread shown (`?show_comments=true`). */
    fun workComments(id: Int): HttpUrl =
        BASE.newBuilder().addPathSegments("works/$id")
            .addQueryParameter("view_adult", "true")
            .addQueryParameter("show_comments", "true")
            .fragment("comments").build()

    /** A specific chapter's page with its comment thread shown (AO3 threads per chapter). */
    fun chapterComments(workId: Int, chapterId: Int): HttpUrl =
        BASE.newBuilder().addPathSegments("works/$workId/chapters/$chapterId")
            .addQueryParameter("view_adult", "true")
            .addQueryParameter("show_comments", "true")
            .fragment("comments").build()

    fun workSubscriptions(id: Int): HttpUrl =
        BASE.newBuilder().addPathSegments("works/$id/subscriptions").build()

    /** The logged-in user's AO3 subscriptions list (works/series/users). */
    fun userSubscriptions(name: String): HttpUrl =
        BASE.newBuilder().addPathSegments("users/$name/subscriptions").build()

    fun mediaFandoms(categoryName: String): HttpUrl =
        "${BASE}media/${ao3PathEncode(categoryName)}/fandoms".toHttpUrl()

    fun tagWorks(tagName: String): HttpUrl =
        "${BASE}tags/${ao3PathEncode(tagName)}/works".toHttpUrl()

    /**
     * Autocomplete is JSON-only and 404s on a raw slash in `term`, so the term
     * is percent-encoded here (including `/` → `%2F`) and set as the raw query.
     */
    fun autocomplete(field: AutocompleteField, term: String): HttpUrl {
        val encoded = percentEncodeTerm(term)
        return "${BASE}autocomplete/${field.path}?term=$encoded".toHttpUrl()
    }

    private fun percentEncodeTerm(term: String): String =
        buildString {
            for (b in term.toByteArray(Charsets.UTF_8)) {
                val c = b.toInt() and 0xFF
                val ch = c.toChar()
                if (ch in 'A'..'Z' || ch in 'a'..'z' || ch in '0'..'9' ||
                    ch == '-' || ch == '_' || ch == '.' || ch == '~'
                ) {
                    append(ch)
                } else {
                    append('%').append("%02X".format(c))
                }
            }
        }

    /** AO3's media-category path encoding (`&`→`*a*`, `/`→`*s*`, `.`→`*d*`, …). */
    private fun ao3PathEncode(s: String): String = buildString {
        for (ch in s) {
            when (ch) {
                '&' -> append("*a*")
                '/' -> append("*s*")
                '.' -> append("*d*")
                '?' -> append("*q*")
                '\'' -> append("*at*")
                ' ' -> append("%20")
                else -> append(ch)
            }
        }
    }
}
